File: solutions/python/03/lower_bound/main.py
import bisect


def lower_bound(a: list, x: int) -> int:
    ng, ok = -1, len(a)
    while ok - ng > 1:
        mid = (ok + ng) // 2
        if a[mid] >= x:
            ok = mid
        else:
            ng = mid
    return ok


def main() -> None:
    a = [1, 3, 3, 5, 7]
    print(bisect.bisect_left(a, 4))
    print(lower_bound(a, 4))


if __name__ == "__main__":
    main()

File: solutions/go/03/lower_bound/main.go
package main

import "fmt"

func lowerBound(a []int, x int) int {
	ng, ok := -1, len(a)
	for ok-ng > 1 {
		mid := (ok + ng) / 2
		if a[mid] >= x {
			ok = mid
		} else {
			ng = mid
		}
	}
	return ok
}

func main() {
	a := []int{1, 3, 3, 5, 7}
	fmt.Println(lowerBound(a, 4))
}
